File: src/javasource/Q4_MergeInterval.java
package javasource;

import java.util.*;
import java.util.stream.Collectors;

public class Q4_MergeInterval {
    public static void main(String[] args) {
        Interval interval1 = new Interval(1, 3);
        Interval interval2 = new Interval(2, 6);

        Interval interval5 = new Interval(7, 9);
        Interval interval3 = new Interval(8, 10);

        Interval interval6 = new Interval(10, 11);

        Interval interval4 = new Interval(15, 18);
        Interval interval7 = new Interval(17, 20);
        Interval interval8 = new Interval(18, 20);

        List<Interval> intervals = Arrays.asList(interval1, interval2, interval3, interval4, interval5, interval6, interval7, interval8);

        Q4_MergeInterval a = new Q4_MergeInterval();
        List<Interval> list = a.merge2(intervals);

        System.out.println(list.toString());
    }

    private List<Interval> merge(List<Interval> intervals) {
        if (intervals.isEmpty()) return intervals;

        intervals.sort((a, b) -> a.start - b.start);
        List<Interval> result = new ArrayList<>();
        result.add(intervals.get(0));
        for(int i = 1; i < intervals.size(); i++) {
            int index = result.size() - 1;
            if (result.get(index).end >= intervals.get(i).start) {
                result.set(index, new Interval(result.get(index).start, intervals.get(i).end));
                i++;
            } else {
                result.add(intervals.get(i));
            }
        }
        return result;
    }

    private List<Interval> merge2(List<Interval> intervals) {
        if (intervals.isEmpty()) return intervals;

        intervals.sort((a, b) -> a.start - b.start);
        List<Interval> result = new ArrayList<>();
        Interval before = intervals.get(0);
        for (int i = 1; i < intervals.size(); i++) {
            Interval current = intervals.get(i);
            if (before.end >= current.start) {
                before.end = current.end;
            } else {
                result.add(before);
                before = current;
            }
        }

        if(!result.contains(before)) {
            result.add(before);
        }

        return result;
    }
}

File: src/kotlinsource/Q4_MergeInterval.kt
package kotlinsource

import java.util.*


fun main(args:Array<String>) {
    val interval1 = Interval(1, 3)
    val interval2 = Interval(2, 6)
    val interval5 = Interval(7, 9)
    val interval3 = Interval(8, 10)
    val interval6 = Interval(10, 11)
    val interval4 = Interval(15, 18)
    val interval7 = Interval(17, 20)
    val interval8 = Interval(18, 20)

    val list:List<Interval> = listOf(interval1, interval2, interval3, interval4, interval5, interval6, interval8, interval7);

    println(merge(list))
}

fun merge(intervals:List<Interval>):List<Interval> {
    if(intervals.isEmpty()) return intervals

    val intervals = intervals.sortedBy { it.start }
    var result = ArrayList<Interval>();
    var before:Interval = intervals.get(0)
    for (i in 1..intervals.size - 1) {
        var current = intervals.get(i)
        if (before.end >= current.start) {
            before.end = current.end;
        } else {
            result.add(before)
            before = current
        }
    }

    if (!result.contains(before)) {
        result.add(before)
    }
    return result
}